text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+support this in the future.
+
+Declaring a range niche with an open start (`..3`) results in an error, for
+forwards-compatibility with support for negative values.
+
+Declaring a niche using a non-literal value (e.g. `usize::MAX`) results in an
+error. Constants can use compile-time evaluation, and compile-time evaluation
+does not occur early enough for attributes such as niche declarations.
+
+If a type `T` contains multiple niche values (e.g. `#[niche(range = 8..16)]`),
+the compiler does not guarantee any particular usage of those niche values in
+the representation of types containing `T`. In particular, the
+compiler does not commit to making use of all the invalid values of the niche,
+even if it otherwise could have.
+
+However, multiple instances of the same identical type (e.g. `Option<T>` and
+`Option<T>`) will use an identical representation (whether the type contains a
+niche or not). This permits a round-trip between such a value and a byte
+representation.
+
+If a type `T` contains niches and uses `repr(C)` or `repr(transparent)`, the
+compiler guarantees to use the same storage size for the type as it would
+without the niche, even if the niche might allow storing fewer bytes. If a type
+`T` contains niches and uses the default (`Rust`) `repr`, the compiler may
+choose to represent the type using fewer bytes if the niche would allow doing
+so. For instance:
+
+```rust
+#[niche(range = 4..)]
+struct S {
+    field: u16,
+}
+
+// `size_of::<S>()` may return less than 2
+```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could allow defining *either* valid or invalid ranges. For instance,
+`niche(invalid_range(0..=3))` or `niche(valid_range(4..))`. Different types
+could use whichever of the two proved simpler for a given use case. However, in
+addition to adding gratuitous complexity and requiring longer names
+(`invalid_range` vs `range`), this would double the number of cases when
+defining other kinds of niches in the future. For instance, a future syntax for
+bit-pattern niches would need to provide both `valid` and `invalid` variants as
+well. We could introduce another level of nesting to make this orthogonal, such
+as `niche(invalid(range(...)))` and `niche(invalid(range(...)))`, but that
+further increases complexity.
+
+Rather than defining the range of *invalid* values, the attribute could define
+the range of *valid* values. Different types may find one or the other case
+simpler. This RFC chooses to define the range of *invalid* values for three
+reasons:
+- As an arbitrary choice, because we need to pick one or the other (see above).
+- The most common case will be a single invalid value, for which defining
+  invalid values results in simpler code.
+- This mechanism commonly goes by the name `niche`, and `niche` also refers to
+  the invalid value. So, an attribute defining the niche of a type most
+  naturally refers to the invalid value.
```

Comment:
I have to say I find the RFC's choice here very confusing. The relevant point for users of this type is that they are refining the validity invariant of their type: they must ensure that all values of this type always are in a certain range. I think it is much more natural to actually state the range you are promising you values to be in, rather than stating the range you are promising your value *not* to be in. That's an unnecessary negation IMO, and bound to lead to confusion.

Niches are just a consequence of the fact that the type has a validity invariant. Honestly I wouldn't even want `niche` in the attribute name, I would expect something like `#[valid_range(...)]`.

Reply:
Also note that in the compiler internally, this is represented as a [range of *valid* values](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/abi/enum.Scalar.html).

Reply:
@RalfJung I feel like that may be a case of viewing the code through formal-methods glasses. :)

In practice, I'm expecting that it's much easier to say `#[niche(value = 0x8000_0000)]` than something like `#[valid_range(0..=0x7FFF_FFFF, 0x8000_0001..=0xFFFF_FFFF)]` or similar. (This is the same motivation I have for providing `value` and not just `range`.)

I can *imagine* cases in which it'd be equally easy, such as the current niche on a "nanoseconds" field, where `0..=999_999_999` and `1_000_000_000..` seem equally easy. But offhand, I have a hard time thinking of a case where it's *easy* to specify the valid values and *hard* to specify the invalid values.

Reply:
Yes that is terrible syntax, but it's not really a fair comparison. The RFC proposes forcing the user to always write their invariant in negated form, so of course if the invariant is negated then that fits nicely. But when the invariant is *not* negated, it becomes awkward.

In contrast, stating the invariant directly of course also gives the user the option to state them in terms of a negation. I was imagining something like `#[valid_range(!= 0x8000_0000)]` or so for that case. We can find syntax to express the invariant "not equal to X". But the key point is that for cases where the invariant is not already negated, such as "value must be in `0..100`", the niche syntax is awkward (you end up with a double negation) but the validity syntax works naturally.

Furthermore, 'niche' is not really a standard term people will have encountered before. The chances are, I think, much higher that people will have encountered the concept of a value always satisfying a certain property (an invariant). Niche is a technical term from the guts of the compiler, and natural extensions of this concept to invariants like `!= 0 && != 10` become a problem because while those are invariants, their negation is not a niche. 

In contrast, the concept of a value being in a certain range of validity is a much higher-level concept that exists not just in Rust but also in other programming languages. Consider for example the ranged integer types from Ada and Pascal. Of course there you state the range the value must be in, not the range the value must *not* be in. (The RFC mentions Ada but not Pascal, and fails to explain why it is a good idea to negate the range when prior art involved stating the range directly.)

I agree that I view things through formal-methods glasses, but that doesn't automatically mean that the formal methods approach is wrong. ;) I expect it will be common to find some unsafe code around types with a custom validity range, and then the programmer has to argue why the validity invariant of the type is maintained. And again, if we phrase things in terms of niches, there will always be an extra negation there. Sometimes a negation is inherent in the invariant ("the value is not `0x8000_0000`"), but sometimes it isn't ("the value is less than 100"). The niche terminology forces users to *always* state their invariant in negated form, which seems natural for "the value is not X" but becomes awkward for invariants that are not already negated.

Reply:
I should also add that thinking in terms of 'niche' might be a case of viewing the code through rustc-implementor glasses. :) I think 'valid range'/'invariant' is a much more common concept, and has much more precedent also outside of formal methods, than 'niche'.

Reply:
Niche fits well for the "single invalid value" case, and range fits rather poorly for that case. It seems reasonable to have both range and individual declaration forms.

Reply:
I'm reopening this because I'm not really convinced it's resolved.  The text now says "The most common case will be a single invalid value", but I couldn't find any justification for that statement.

Near as I can tell, the vast majority of types using the attribute in the compiler today use multiple invalid values, via the default range of `0..=0xFFFF_FF00` in `rustc_index::newtype_index!`:

https://github.com/rust-lang/rust/blob/94b2b15e63c5d2b2a6a0910e3dae554ce9415bf9/compiler/rustc_macros/src/newtype.rs#L129-L130

Or looking at the library, `Nanoseconds` jumps out

https://github.com/rust-lang/rust/blob/1481fd964bac3c750c7e1b21206fdaa60346c456/library/core/src/time.rs#L34-L36

both as something with lots of invalid values *and* where saying "the valid range is `0..=999_999_999`" feels *far* more natural to me than saying "the invalid range is `1_000_000_000..=4_294_967_295`".

This feels pretty broadly applicable to me, in fact.  I would describe the ASCII codepoints as those in `0..=127`, for example.  I wouldn't call them the code points that *aren't* in `128..=0x10FFFF`.  Even for things that aren't ranges, they're normally described positively.  For example, LLVM [says](https://llvm.org/docs/LangRef.html#global-variables) "The maximum alignment is `1 << 32`", not "alignments of `1<<33` and higher are disallowed".

So if single holes are the only place where the negative phrasing is more natural (and are holes other than zero, which doesn't need this RFC, and `!0` all that common?), that says to me we just should have a nice way to convert from a single hole to the valid rather.  Which is another thing that pushes me to not wanting this to be an attribute, because we have a way to define names for simple cases that can be implemented in terms of a more-complex form:
```
pub type U32WithHole<const H: u32> = RefinedU32<{H.wrapping_add(1)}, {H.wrapping_sub(1)}>;
```

Reply:
> Niche fits well for the "single invalid value" case, and range fits rather poorly for that case. It seems reasonable to have both range and individual declaration forms.

I should probably have said 'set', not 'range'. The "single invalid value" case should, in my opinion, conceptually be declared as "all values are valid except for this one". I agree it should not be defined as a range with lower and upper bounds.

Reply:
The first use case I could think of for a user-defined niche would be risc-v registers, which I implemented with the valid range 1..=31 (so, nonzero _and_ less than 32). Just as a datapoint for "easy to specify valid values, hard to specify invalid"